File: fused_render/shell/recents.py
"""GET/PUT/POST /api/recents — recently opened files at ~/.fused-render/recents.json.

The sidebar's "Recents" section: the last files opened in the shell, each with
the last params they carried (the entry url is updated live while the file is
open — see frontend lib/recents.ts). Files only: directory navigation and
sentinel routes (`_panel`, `_prefs`, any `_`-prefixed view) are never recorded.

File shape:

    {"collapsed": false, "entries": [
        {"url": "/view/...?...", "openedAt": iso, "title": "optional page title"}
    ]}

Entries are newest-first, deduped by target fs path, capped at ENTRY_CAP (a
buffer so the UI's top 3 survive missing-file filtering). URLs are stored
verbatim including the query string — same "URL is the whole state" posture as
bookmarks (D20). `collapsed` lives in the data file itself, matching D44's
persisted folder collapse. Entries whose file has since been deleted are
hidden from the GET response but never deleted from disk — the file may come
back (a mount reconnect, an undeleted trash item).
"""
import asyncio
import os
from concurrent.futures import ThreadPoolExecutor
from datetime import datetime, timezone
from urllib.parse import unquote, urlsplit

from fastapi import APIRouter, Body, Header
from fastapi.responses import JSONResponse

from fused_render.shell import storage

router = APIRouter()

VIEW_PREFIX = "/view/"

# Cap well above the UI's 3 visible rows so enough valid entries remain after
# missing-file filtering.
ENTRY_CAP = 20

# Total wall-clock budget for ALL of a GET /api/recents' existence checks. The
# checks fan out concurrently (on _CHECK_POOL below) and share this one deadline
# — so the endpoint stays bounded no matter how many entries sit on a slow or hung
# mount. An entry whose check outlives the budget is KEPT, not dropped (fail
# open): a possibly-dead sidebar row beats a multi-second stall. Sized like the
# fs/events watch ticker's per-stat _STAT_TIMEOUT_S (server.py, 4.0s) but tighter
# — this is the whole sidebar section, not one background poll.
CHECK_BUDGET_S = 1.5

# Existence checks run on a DEDICATED pool, not asyncio's default loop executor:
# a check that outlives CHECK_BUDGET_S keeps running in its thread (a blocking
# os.stat/rc call can't be cancelled mid-flight), and parking those on the
# shared default executor makes the request wait for them at loop teardown.
# ENTRY_CAP-wide so every entry's check can run at once (true fan-out); slots
# free themselves as the local isfile / bounded rc_mtime_for calls return.
_CHECK_POOL = ThreadPoolExecutor(
    max_workers=ENTRY_CAP, thread_name_prefix="recents-exists"
)


def _require_fused(x_fused: str | None) -> JSONResponse | None:
    # Same D3 guard as server._require_fused, duplicated to keep shell↛server
    # acyclic (see shell/bookmarks.py).
    if x_fused != "1":
        return JSONResponse({"error": "missing X-Fused header"}, status_code=403)
    return None


def _path() -> str:
    return os.path.join(storage.home_dir(), "recents.json")


def _decoded_fs_path(url: str) -> str | None:
    """Decode a shell url to the absolute fs path it names, or None when it
    cannot name a file at all: non-/view/ urls and sentinel routes (any
    `_`-prefixed top-level pathname — `_panel`, `_prefs`, `_templates`, ...).

    Deliberately does NOT check disk: this is the dedupe identity, and two
    entries for the same path must dedupe even while the file is deleted
    (existence matters only for GET filtering and for accepting a new record —
    see _file_path_from_url).

    Mirrors bookmarks._fs_path_from_url but is stricter: only under /view/
    (recording never happens in embed panes)."""
    try:
        parts = urlsplit(url)
    except ValueError:
        return None
    path = parts.path
    if not path.startswith(VIEW_PREFIX):
        return None
    segments = [unquote(s) for s in path[len(VIEW_PREFIX):].split("/") if s]
    # Any `_`-prefixed top-level pathname is a shell sentinel (`_panel`,
    # `_prefs`, `_templates`, `_listing`, ...) — the whole namespace is
    # shell-owned, so reject the prefix rather than enumerating names.
    if not segments or (len(segments) == 1 and segments[0].startswith("_")):
        return None
    joined = "/".join(segments)
    # Mirror the frontend's rootedFsPath (lib/router.ts): keep a Windows
    # drive-letter path absolute as-is; prepend "/" only for POSIX paths.
    if len(joined) == 2 and joined[0].isalpha() and joined[1] == ":":
        fs_path = joined + "/"
    elif len(joined) >= 3 and joined[0].isalpha() and joined[1] == ":" and joined[2] == "/":
        fs_path = joined
    else:
        fs_path = "/" + joined
    return fs_path


def _file_path_from_url(url: str) -> str | None:
    """`_decoded_fs_path`, additionally requiring the path to be an existing
    FILE right now (recents record files only, D22-style basename rows) —
    the gate for accepting a record and for the GET response filter."""
    fs_path = _decoded_fs_path(url)
    if fs_path is None:
        return None
    # NEVER a raw os.path.isfile on a mount-backed path: a cold GETATTR there
    # lists the whole parent prefix and wedges the mount (the open-flow wedge —
    # POST /api/recents/open resolves the just-opened file through here). Route
    # mounts via the rc API (_mount_exists), locals via the kernel (_local_exists).
    from fused_render.shell import mounts as shell_mounts
    exists = (_mount_exists(fs_path) if shell_mounts.is_mount_backed(fs_path)
              else _local_exists(fs_path))
    return fs_path if exists else None


def _local_exists(fs_path: str) -> bool:
    """Existence of a LOCAL (non-mount-backed) path. A plain os.path.isfile is
    safe and cheap here — the mount-wedging GETATTR concern only applies under a
    managed mount, which _keep_entry routes away from this call."""
    try:
        return os.path.isfile(fs_path)
    except OSError:
        return False


def _mount_exists(fs_path: str) -> bool:
    """Whether a MOUNT-BACKED path is an existing FILE, answered by the rclone
    rc API (mounts.rc_kind_for) — NEVER os.path.isfile. A raw os.stat/isfile on
    a hung NFS mount is the exact GETATTR that wedges it (see rc_kind_for/
    rc_mtime_for in mounts.py); the rc route keeps the kernel out of the loop
    entirely.

    Files only, matching recents' D22 files-only contract (and _local_exists'
    os.path.isfile): a confirmed "dir" filters the entry just like a "missing"
    would. Only a "file" or an "indeterminate" probe (rcd down / timed out /
    errored — rc can't prove anything) keeps it: we fail open rather than hide a
    live recent on a transient rc hiccup."""
    from fused_render.shell import mounts as shell_mounts

    try:
        return shell_mounts.rc_kind_for(fs_path) in ("file", "indeterminate")
    except Exception:
        return True  # unexpected error -> fail open, keep the entry


async def _keep_entry(url: str) -> bool:
    """Whether a recents entry should appear in the GET response: True to keep,
    False to filter (file confirmed gone). Runs the existence check off the
    event loop and routes mount-backed paths through the rc API. Only a check
    that COMPLETES False filters; anything indeterminate keeps (fail open)."""
    fs_path = _decoded_fs_path(url)
    if fs_path is None:
        return False  # not a file-naming url (sentinel / non-/view/) -> filtered
    from fused_render.shell import mounts as shell_mounts

    check = _mount_exists if shell_mounts.is_mount_backed(fs_path) else _local_exists
    loop = asyncio.get_running_loop()
    try:
        return await loop.run_in_executor(_CHECK_POOL, check, fs_path)
    except Exception:
        return True  # fail open on any unexpected error


def _read() -> dict:
    data = storage.read_json(_path())
    if not isinstance(data, dict):
        return {"collapsed": False, "entries": []}
    entries = data.get("entries")
    return {
        "collapsed": data.get("collapsed") is True,
        "entries": [e for e in entries if isinstance(e, dict)] if isinstance(entries, list) else [],
    }


@router.get("/api/recents")
async def get_recents():
    """The collapsed flag + entries whose target file still exists. Missing
    files are filtered from the response, NOT deleted from disk (read-only
    GET, like bookmarks; the file may reappear).

    Existence checks fan out concurrently under a single CHECK_BUDGET_S deadline
    and are mount-safe (mount-backed paths route through the rclone rc API, not
    a kernel os.stat). An entry whose check outlives the budget is KEPT (fail
    open) — a possibly-dead row beats a stalled sidebar."""
    data = _read()
    raw = [e for e in data["entries"] if isinstance(e.get("url"), str)]
    keeps = [True] * len(raw)  # default keep (fail open on timeout/cancel)
    if raw:
        tasks = [asyncio.ensure_future(_keep_entry(e["url"])) for e in raw]
        done, pending = await asyncio.wait(tasks, timeout=CHECK_BUDGET_S)
        for i, t in enumerate(tasks):
            if t in done:
                # _keep_entry never raises, but stay defensive: keep on error.
                keeps[i] = t.result() if not t.cancelled() else True
        for t in pending:
            t.cancel()  # deadline hit; underlying thread finishes on its own
    entries = [e for e, keep in zip(raw, keeps) if keep]
    return {"collapsed": data["collapsed"], "entries": entries}


@router.post("/api/recents/open")
def post_recent_open(
    payload: dict = Body(...), x_fused: str | None = Header(default=None)
):
    guard = _require_fused(x_fused)
    if guard is not None:
        return guard
    url = payload.get("url")
    if not isinstance(url, str):
        return JSONResponse({"error": "url required"}, status_code=400)
    fs_path = _file_path_from_url(url)
    if fs_path is None:
        # Directory / sentinel / missing-file url -> benign no-op, same
        # posture as POST /api/bookmarks/history for non-file urls.
        return {"recorded": False}
    # The rendered page's own <title> (frontend lib/recents.ts), preferred
    # over the file's basename by the sidebar row — same posture as bookmark
    # naming. Optional: absent until the preview iframe reports one.
    title_raw = payload.get("title")
    title = title_raw.strip() if isinstance(title_raw, str) and title_raw.strip() else None
    data = _read()
    # Dedupe by DECODED target fs path — existence-blind, so a dead entry for
    # the same path (file deleted and recreated since) is replaced rather than
    # left wasting a cap slot beside the fresh one. A title-less record (the
    # synchronous open-record always fires before the async title arrives;
    # some modes never report one at all) must not erase a title learned on a
    # PREVIOUS open — carry the old one forward when this record doesn't
    # supply a fresher one, so a known title only ever improves, never regresses.
    existing_title = None
    kept = []
    for e in data["entries"]:
        if isinstance(e.get("url"), str) and _decoded_fs_path(e["url"]) == fs_path:
            t = e.get("title")
            if existing_title is None and isinstance(t, str) and t:
                existing_title = t
            continue
        kept.append(e)
    entry = {"url": url, "openedAt": datetime.now(timezone.utc).isoformat()}
    if title is not None:
        entry["title"] = title
    elif existing_title is not None:
        entry["title"] = existing_title
    data["entries"] = [entry, *kept][:ENTRY_CAP]
    storage.write_json(_path(), data)
    return {"recorded": True}


@router.put("/api/recents/collapsed")
def put_recents_collapsed(
    payload: dict = Body(...), x_fused: str | None = Header(default=None)
):
    guard = _require_fused(x_fused)
    if guard is not None:
        return guard
    collapsed = payload.get("collapsed")
    if not isinstance(collapsed, bool):
        return JSONResponse({"error": "'collapsed' must be a boolean"}, status_code=400)
    data = _read()
    data["collapsed"] = collapsed
    storage.write_json(_path(), data)
    return {"collapsed": collapsed}

File: frontend/src/apps/explorer/lib/recents.ts
// Recents store + tracking hook, persisted server-side at
// ~/.fused-render/recents.json via /api/recents (fused_render/shell/recents.py).
//
// Reads are synchronous off an in-memory cache (same posture as bookmarks.ts);
// the server owns all list logic — dedupe by fs path, newest-first order, the
// 20-entry cap, missing-file filtering — so every mutation here is a POST/PUT
// followed by a cache refresh from the response of a fresh GET. Recording is
// fire-and-forget: a recents failure must never affect the view being opened.
import { useEffect, useRef } from "react";

import { getRecents, postRecentOpen, putRecentsCollapsed } from "@platform/lib/api";
import type { RecentEntry, RecentsResult } from "@platform/lib/api";
import { useEventCounter } from "@platform/lib/hooks";
import { stripSessionParams } from "@platform/lib/session-params";
import { IS_EMBED, VIEW_PREFIX, currentUrl, fsPathFromLocation, rootedFsPath } from "@platform/lib/router";

export type { RecentEntry };

// Store change signal — explorer-owned (recents are an explorer concept, not a
// shell/platform one; the app-builder keeps its own independent recents in
// apps/builder/lib/recents.ts). The store dispatches it itself after every
// cache advance.
const RECENTS_EVENT = "fused:recents";

export function notifyRecentsChanged(): void {
  window.dispatchEvent(new Event(RECENTS_EVENT));
}

export function useRecentsVersion(): number {
  return useEventCounter([RECENTS_EVENT]);
}

let cache: RecentsResult = { collapsed: false, entries: [] };

export function loadRecents(): RecentsResult {
  return cache;
}

// The fs path a recent entry targets, decoded from its explorer url — the
// entry's stable identity: the url mutates on every live param write, the
// path doesn't (React row keys and the slot order below key on it). The bare
// legacy "/view/" prefix still decodes (entries recorded before the /explorer
// namespace rename).
export function recentFsPath(url: string): string {
  const qIdx = url.indexOf("?");
  const pathname = qIdx !== -1 ? url.slice(0, qIdx) : url;
  const prefix = [VIEW_PREFIX, "/view/"].find((p) => pathname.startsWith(p));
  if (!prefix) return pathname;
  return rootedFsPath(
    pathname.slice(prefix.length).split("/").filter(Boolean).map(decodeURIComponent).join("/")
  );
}

// --- stable-slot display order ----------------------------------------------
//
// The DATA is strict MRU (the server moves a re-recorded file to the top),
// but displaying raw MRU makes the list jump under the user's own pointer:
// clicking a shown recent, or param churn on the open file, would reshuffle
// rows mid-interaction. So the visible top-3 uses session-scoped stable
// slots: a displayed file keeps its slot for the whole page session — its
// row just updates in place — and the only movement is a file NOT currently
// displayed entering at the top (a real navigation), pushing the bottom row
// out. A displayed file that vanishes (deleted; GET filters it) leaves its
// slot and the next MRU entry fills in at the BOTTOM — survivors never
// reshuffle. Not persisted: on boot the slots seed from server MRU order.

const DISPLAY_ROWS = 3;

let slotPaths: string[] = [];

function computeSlots(prev: string[], entries: RecentEntry[]): string[] {
  const mruPaths = entries.map((e) => recentFsPath(e.url));
  const alive = new Set(mruPaths);
  // Vanished files leave their slot; survivors keep their relative order.
  let slots = prev.filter((p) => alive.has(p));
  // A file not currently displayed entering at the MRU head is a real new
  // open -> the one allowed movement: insert at top, bottom row falls out.
  const head = mruPaths[0];
  if (head !== undefined && !slots.includes(head)) slots = [head, ...slots];
  // Fill any remaining vacancies from the bottom, in MRU order.
  for (const p of mruPaths) {
    if (slots.length >= DISPLAY_ROWS) break;
    if (!slots.includes(p)) slots.push(p);
  }
  return slots.slice(0, DISPLAY_ROWS);
}

// The entries to display, in stable-slot order (each slot carries its file's
// LATEST entry — url updates land in place). Idempotent per cache state, so
// safe to call on every sidebar render.
export function displayRecents(): RecentEntry[] {
  slotPaths = computeSlots(slotPaths, cache.entries);
  const byPath = new Map(cache.entries.map((e) => [recentFsPath(e.url), e]));
  return slotPaths.flatMap((p) => byPath.get(p) ?? []);
}

// Drop the rows for a path the user just DELETED (trashed or hard-deleted), and for
// everything inside it when that path was a folder — the same prefix + "/"
// containment test `clearClipboardIfDeleted` uses.
//
// The GET already hides entries whose file is gone (RC-7), but nothing re-GETs
// after a delete: the row would sit in the sidebar pointing at a file that no
// longer exists until the user's next navigation happened to refresh the cache.
// So the delete tells us, and the cache drops it right there.
//
// Deliberately LOCAL — no request. The server's list runs deeper than the three
// displayed rows (RC-6's 20-entry buffer exists for exactly this), so the vacated
// slot refills from the cache we already hold, with no round-trip and nothing to
// wait for. A re-GET would also be the wrong tool: RC-7's existence check fails
// OPEN on an indeterminate answer (rc down, budget exceeded), which for a row the
// user just deleted would mean watching it come back. Nothing is written to the
// store — the entry stays on disk, hidden, exactly as RC-7 has it, so a file
// restored from the Bin legitimately reappears in Recents.
export function dropRecentsFor(deleted: string): void {
  const kept = cache.entries.filter((e) => {
    const path = recentFsPath(e.url);
    return path !== deleted && !path.startsWith(deleted + "/");
  });
  if (kept.length === cache.entries.length) return;
  cache = { ...cache, entries: kept };
  // Same slot arithmetic a vanished-on-refresh entry gets (RC-11): the freed slot
  // is filled from the bottom by the next MRU entry, survivors do not reshuffle.
  slotPaths = computeSlots(slotPaths, cache.entries);
  notifyRecentsChanged();
}

// Serial promise chain like bookmarks.ts's enqueue: recording bursts (open +
// the debounced param updates) and the collapse toggle never interleave their
// GET-after-write refreshes, so the cache can't step backwards to a stale read.
let tail: Promise<unknown> = Promise.resolve();

function enqueue<T>(op: () => Promise<T>): Promise<T> {
  const run = tail.then(op, op);
  tail = run.catch(() => {});
  return run;
}

// What the sidebar actually renders from this store: the collapse flag plus
// each displayed slot's (path, latest url, latest title). Urls are INCLUDED —
// a stale href is a real bug (middle-click/copy-link navigates to outdated
// params, RC-3); with stable slots + path-keyed rows a url-only notify
// re-renders the anchor attributes in place with zero movement and zero
// remounts. Title is included for the same reason: the open-record and the
// title-report re-record that follows it share the same url (only the title
// differs), so a url-only signature would miss that change entirely and the
// row would keep showing the basename after the cache already has the title.
function displaySignature(slots: string[], entries: RecentEntry[], collapsed: boolean): string {
  const byPath = new Map(entries.map((e) => [recentFsPath(e.url), e]));
  return JSON.stringify([
    collapsed,
    slots.map((p) => {
      const e = byPath.get(p);
      return [p, e?.url, e?.title];
    }),
  ]);
}

async function refresh(): Promise<void> {
  const prevSig = displaySignature(slotPaths, cache.entries, cache.collapsed);
  cache = await getRecents();
  // A snapshot fetched while a collapse write is in flight predates that
  // write — the user's newest intent stays authoritative over it.
  if (pendingCollapsed !== null) cache = { ...cache, collapsed: pendingCollapsed };
  slotPaths = computeSlots(slotPaths, cache.entries);
  // Notify only when the visible slice changed; identical-signature refreshes
  // (e.g. a re-record of an unchanged url) stay render-free.
  if (displaySignature(slotPaths, cache.entries, cache.collapsed) !== prevSig) {
    notifyRecentsChanged();
  }
}

// Load the cache once at boot (main.tsx, beside hydrateBookmarks).
export function hydrateRecents(): Promise<void> {
  return enqueue(() =>
    refresh().catch((e) => console.error("[fused] failed to load recents:", e))
  );
}

// SESSION-ONLY PARAMS ARE STRIPPED FIRST (LSN-12, D326). Recents is an AUTOMATIC
// capture that lands on disk: this hook fires on every `fused:urlchange`, so
// without this a click that CLOSES the file preview's companion sidebar
// (`?_side=off`) is recorded, and every later open from the Recents list comes up
// shut — the sidebar's state persisted for good, by a write the user never asked
// for, which is the exact thing D326 exists to stop. The URL a recent replays must
// hold what the file WAS, not what the chrome around it was doing.
//
// Deliberately NOT applied to BOOKMARKS: a bookmark is an explicit "save this
// view" gesture and SB-2 says it captures the URL verbatim, which is how `_mode`,
// sort and a chosen `_side` companion all end up in one. Same param, opposite
// answer, because one capture is chosen and the other is a side effect.
function stripRecordedParams(url: string): string {
  const q = url.indexOf("?");
  if (q < 0) return url;
  const kept = stripSessionParams(url.slice(q + 1));
  return kept === "" ? url.slice(0, q) : url.slice(0, q + 1) + kept;
}

// Record an open (or a live param update) of the current file view. The
// server dedupes by target fs path — a re-record of an already-listed file
// moves it to the top and replaces its url — and no-ops for anything that is
// not an existing file's /view/ url, so the caller stays dumb about the
// target's kind. `title`, when known (the page's own <title>, see App.tsx's
// StatView), is stored alongside the url so the sidebar row can prefer it
// over the file's basename — same posture as bookmark naming.
export function recordRecentOpen(url: string, title?: string | null): Promise<void> {
  return enqueue(async () => {
    try {
      await postRecentOpen(stripRecordedParams(url), title);
      await refresh();
    } catch (e) {
      console.error("[fused] failed to record recent open:", e);
    }
  });
}

// The user's latest collapse intent, authoritative while any write is in
// flight: a refresh() snapshot fetched before the PUT landed must not undo
// the optimistic flip, and under rapid toggles only the NEWEST intent may
// win — an older call never re-asserts its own stale value on completion.
let pendingCollapsed: boolean | null = null;
let collapseWritesInFlight = 0;

export function setRecentsCollapsed(collapsed: boolean): Promise<void> {
  // Optimistic: the flip paints NOW, not a network round-trip later — the
  // collapsed rail's Recents icon expands the sidebar in the same click and
  // must reveal the list it promised, and the heading toggle should not lag
  // the pointer either. Persistence follows behind, serialized by enqueue.
  pendingCollapsed = collapsed;
  collapseWritesInFlight++;
  cache = { ...cache, collapsed };
  notifyRecentsChanged();
  return enqueue(async () => {
    let persisted = true;
    try {
      await putRecentsCollapsed(collapsed);
    } catch (e) {
      persisted = false;
      console.error("[fused] failed to persist recents collapse:", e);
    }
    if (--collapseWritesInFlight > 0) return; // a newer intent is still writing
    pendingCollapsed = null;
    // The cache already shows the latest intent; a successful FINAL write
    // just confirmed the server agrees. A failed final write means the
    // server kept some earlier state — reload it rather than lie about what
    // survives a refresh.
    if (!persisted) {
      await refresh().catch((e) => console.error("[fused] failed to reload recents:", e));
    }
  });
}

// Track-on-open + live param updates. Mounted by StatView: records once when
// the stat confirms a file, then re-records the current url on every param
// write (fused:urlchange — the iframe runtime's replaceState is wrapped in
// main.tsx) with a 500 ms debounce against slider-style param churn. Embed
// panes, directories, and not-yet-stat'd opens (isDir null) opt out; the server
// rejects non-file urls anyway. This confirmed-file gate is the last survivor
// of the seam it used to share with the per-file session restore's tracking
// hook (lib/session.ts, removed in D329).
// `title` is the previewed page's own <title>, when known — it arrives async
// (after the iframe loads), so it is also a dependency: once it resolves, the
// effect re-runs and re-records the current url with the now-known title,
// same as a live param update would.
export function useRecentsTracking(fsPath: string, isDir: boolean | null, title: string | null): void {
  const timer = useRef<number | undefined>(undefined);
  useEffect(() => {
    if (IS_EMBED || isDir !== false) return;
    // The url waiting out the debounce. Captured at EVENT time, while the
    // shell still shows this file — the unmount flush below runs after the
    // location has already moved on, so reading currentUrl() there would
    // either drop the update or record the next view's url.
    let pending: string | null = null;
    const flush = () => {
      window.clearTimeout(timer.current);
      if (pending !== null) {
        void recordRecentOpen(pending, title);
        pending = null;
      }
    };
    // The open itself (session restore's replaceState re-records with the
    // restored params). Guarded against a same-tick navigation race.
    if (fsPathFromLocation() === fsPath) void recordRecentOpen(currentUrl(), title);
    const onUrlChange = () => {
      if (fsPathFromLocation() !== fsPath) return; // navigated away — not ours
      pending = currentUrl();
      window.clearTimeout(timer.current);
      timer.current = window.setTimeout(flush, 500);
    };
    window.addEventListener("fused:urlchange", onUrlChange);
    window.addEventListener("popstate", onUrlChange);
    return () => {
      window.removeEventListener("fused:urlchange", onUrlChange);
      window.removeEventListener("popstate", onUrlChange);
      // Navigating away inside the debounce window must not lose the last
      // param state — flush the captured url instead of dropping it.
      flush();
    };
    // fsPath + isDir identify the open; title is included so its late arrival
    // triggers a re-record.
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [fsPath, isDir, title]);
}
